src/lib.rs

```diff
@@ -98,43 +120,28 @@ where
         Err(Error::Ack)
     }
 
-    fn start(&mut self) {
-        self.clk.set_low();
-        self.dio.set_high();
-        self.clk.set_high();
+    fn start(&mut self) -> Result<(), Error> {
+        self.clk.set_low().map_err(|_| Error::IO).unwrap();
+        self.dio.set_high().map_err(|_| Error::IO).unwrap();
+        self.clk.set_high().map_err(|_| Error::IO).unwrap();
```

Comment:
Duplicating `.map_err(|_| Error::IO).unwrap()` doesn't seem to increase the binary size, but, violates the DRY principle. However, I don't see a workaround yet.

I tried to wrap the original error into IO by making the Error a generic:
```
pub enum Error<E> {
    Ack,
    IO(E),
}
```
But, it cost me 16 changed lines of `map_err` code.

Reply:
I don't know of a way around this, it's sort of just an implementation detail. If a pin state change somehow fails, that results in an error type from within `embedded-hal` and we need to convert it into our defined error type. Not mapping the error types results in a bunch of unused `Result` types, which causes compiler warnings.

Reply:
I believe lack of mapping will cause a different behavior - it won't stop the statement on first error (just like it was with v1).
My thinking is to put blocks of pin-switch calls into a separate functions and map the function results rather than mapping each particular pin-switch.
Also, it would be better to delegate the error handling to the library user (just pass/wrap the original errors somehow). So, the lib wouldn't mask/lose error details. But, it may be a matter of a different PR/fix.

Reply:
Figured out the proper way to do it - https://doc.rust-lang.org/stable/rust-by-example/error/multiple_error_types/wrap_error.html
```rust
#[derive(Debug)]
pub enum Error<E> {
    Ack,
    IO(E),
}

impl<E> From<E> for Error<E> {
    fn from(err: E) -> Error<E> {
        Error::IO(err)
    }
}
```
does all the trick.

Reply:
Yay, I managed to push into your MR :smile:

Reply:
Well, I'm ready to merge this.
@jessebraham please, take a look and lemme know if you're OK and the code fits your expectations.

Reply:
Doesn't matter to me, it's your library. I just want it to actually work and I'm happy.

Reply:
That's why I'm asking)
I checked it works on the stm32f103, but, I'm not sure if it works on your hardware. And if it suits your API calls.

Ok, then it will be merged/published soon. Also, I figured out some way to reduce the compiled code size even more (which is good for weak MCUs).
BTW, what hardware/SoC/MCU do you use?